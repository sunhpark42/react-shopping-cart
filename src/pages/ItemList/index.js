import React from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { addItemToCart } from '../../store/cartReducer';
import { API } from '../../utils';
import { MESSAGE, PATH } from '../../constants';
import { Grid, Card, IconButton } from '../../components/shared';
import { Description, Price } from './style';
import { ReactComponent as CartIcon } from '../../assets/icons/cart.svg';

const ItemList = () => {
  const list = useSelector(state => state.reducer.itemList);
  const dispatch = useDispatch();
  const history = useHistory();

  const goDetailPage = id => () => {
    history.push(`${PATH.GOODS_DETAIL}?id=${id}`);
  };

  const addCart = id => async () => {
    try {
      const newCartItem = await API.addItemToCart(id);

      dispatch(addItemToCart(newCartItem));
      alert(MESSAGE.SUCCESS_ADD_ITEM_TO_CART);
    } catch (error) {
      console.error(error.message);
      alert(MESSAGE.FAIL_ADD_ITEM_TO_CART);
    }
  };

  return (
    <Grid col="4">
      {Object.values(list).map(({ product_id: id, name, image_url, price }) => {
        return (
          <Card
            key={id}
            title={name}
            thumbnail={{ image: image_url, alt: name }}
            onClick={goDetailPage(id)}
            description={
              <Description>
                <Price>{price.toLocaleString('ko-KR')} 원</Price>
                <IconButton
                  size="medium"
                  ariaLabel={`${name}을 장바구니에 담기`}
                  onClick={addCart(id)}
                >
                  <CartIcon />
                </IconButton>
              </Description>
            }
          />
        );
      })}
    </Grid>
  );
};

export default ItemList;
