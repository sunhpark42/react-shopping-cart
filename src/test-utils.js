// test-utils.js
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer

import { reducer } from './store';
import { cartReducer } from './store/cartReducer';

const rootReducer = combineReducers({ cartReducer, reducer });

function render(
  ui,
  { initialState, store = createStore(rootReducer, initialState), ...renderOptions } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
