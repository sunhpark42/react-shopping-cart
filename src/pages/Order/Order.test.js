import React from 'react';
import ItemList from '.';
import { render, fireEvent, waitFor } from '../../test-utils';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';

const initialState = {
  cartReducer: {
    cart: [
      {
        id: 1,
        name: '[든든] 유부 슬라이스 500g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png',
        price: 4900,
        quantity: 4,
        checked: true,
      },
    ],
  },
  reducer: {
    itemList: [
      {
        id: 1,
        name: '[든든] 유부 슬라이스 500g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png',
        price: 4900,
      },
      {
        id: 2,
        name: '[든든] 진맛살 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11157-main-01.png',
        price: 7100,
      },
      {
        id: 3,
        name: '[든든] 신-키리모찌 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11161-main-01.png',
        price: 12000,
      },
      {
        id: 4,
        name: '[든든] 흑곤약 250g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11263-main-01.png',
        price: 1300,
      },
      {
        id: 5,
        name: '[든든] 흑곤약 1.2kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11277-main-01.png',
        price: 4000,
      },
      {
        id: 6,
        name: '[든든] 곤약말이 시라타끼 200g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11280-main-01.png',
        price: 1100,
      },
      {
        id: 7,
        name: '[든든] 기분 시로가마보코 160g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11315-main-01.png',
        price: 2800,
      },
      {
        id: 8,
        name: '[든든] 냉동조미유부삼각 (60장) 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11323-main-01.png',
        price: 8400,
      },
      {
        id: 9,
        name: '[든든] 시노다마끼 유부야채말이 (15g*40입) 600g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11328-main-01.png',
        price: 9000,
      },
      {
        id: 10,
        name: '[든든] 기분 오뎅세트 433g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11331-main-01.png',
        price: 8900,
      },
    ],
    orderList: [],
  },
};

describe('상품목록 페이지', () => {
  it('store에 있는 products가 제대로 렌더링 되는지 확인한다.', () => {
    const utils = render(<ItemList />, { initialState });

    expect(utils.container).toMatchSnapshot();
  });

  it('결제 버튼 클릭 시 주문하시겠습니까? 창이 나타나고, 확인을 누를시 주문 정보 페이지로 이동한다.', async () => {
    const utils = render(
      <Router>
        <ItemList />
      </Router>,
      { initialState },
    );
    const purchaseButton = utils.getByRole('button', {
      name: /19,600원 주문하기/i,
    });
    const alertSpy = jest.spyOn(window, 'alert');
    alertSpy.mockImplementation(jest.fn(() => true));

    fireEvent.click(purchaseButton);

    await waitFor(() => expect(window.location.pathname).toBe('/mymart/order/detail'));
  });
});
