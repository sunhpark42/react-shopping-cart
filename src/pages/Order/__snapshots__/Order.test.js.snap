// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`상품목록 페이지 store에 있는 products가 제대로 렌더링 되는지 확인한다. 1`] = `
<div>
  <div
    class="sc-ieebsP cLKEWA"
  >
    <h2
      class="sc-dJjZJu fqvgKC"
    >
      주문/결제
    </h2>
    <div
      class="sc-hGPAah jgGCxB"
    >
      <div
        class="sc-dlVyqM bLAYPQ"
      >
        <div
          class="sc-kfPtPH"
        >
          <h3
            class="sc-fKVsgm jzfvKM"
          >
            주문 상품(
            1
            건)
          </h3>
          <div
            class="sc-bBHwJV"
          >
            <div
              class="sc-iwjezw gHlSBW"
            >
              <div
                class="sc-bqiQRQ eEanMH"
              >
                <div
                  class="sc-gsDJrp"
                >
                  <div
                    class="sc-dkPtyc fPszsp"
                  >
                    <img
                      alt="[든든] 유부 슬라이스 500g"
                      class="sc-hKwCoD doJTuW"
                      src="https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png"
                    />
                  </div>
                </div>
                <div
                  class="sc-ksdxAp jfWVEg"
                >
                  <h4
                    class="sc-hBURRC hWUoEx"
                  >
                    [든든] 유부 슬라이스 500g
                  </h4>
                  <div
                    class="sc-fotPbf jqZEHt"
                  >
                    수량: 4
                  </div>
                </div>
                <div
                  class="sc-fFehDp bItAXM"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-cxpRKc ecubRC"
      >
        <h3
          class="sc-llYToB cyPWaB"
        >
          결제금액
        </h3>
        <div
          class="sc-iJKOzS kAsGyE"
        >
          <div
            class="sc-giYgFv GpuDQ"
          >
            <span
              class="sc-pVTma ixgaok"
              color="#94E0DD"
              font-size="1.25rem"
            >
              총 결제금액
            </span>
            <span
              class="sc-pVTma ixgaok"
              color="#94E0DD"
              font-size="1.25rem"
            >
              19,600 원
            </span>
          </div>
          <button
            class="sc-furvIG fkHJhn"
            color="#FFFFFF"
            type="button"
          >
            19,600원 주문하기
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
